src/lib/group.ts

```diff
@@ -0,0 +1,364 @@
+import { Scalar, withMessage } from './core.js';
+import { Field, FieldVar, isField } from './field.js';
+import { Bool, Snarky } from '../snarky.js';
+import { Field as Fp } from '../provable/field-bigint.js';
+import { Pallas } from '../bindings/crypto/elliptic_curve.js';
+
+export { Group };
+
+type FieldLike = FieldVar | Field | number | string | bigint;
+type GroupLike = {
+  x: FieldLike;
+  y: FieldLike;
+};
+
+/**
+ * An element of a Group.
+ */
+class Group {
+  x: Field;
+  y: Field;
+
+  /**
+   * The generator `g` of the Group.
+   */
+  static get generator() {
+    return new Group({ x: Pallas.one.x, y: Pallas.one.y });
+  }
+
+  /**
+   * Coerces anything group-like to a {@link Group}.
+   */
+  constructor({
+    x,
+    y,
+  }: {
+    x: FieldVar | Field | number | string | bigint;
+    y: FieldVar | Field | number | string | bigint;
+  }) {
+    this.x = isField(x) ? x : new Field(x);
+    this.y = isField(y) ? y : new Field(y);
+
+    if (this.#isConstant()) {
+      const { add, mul, square } = Fp;
+
+      let x_bigint = this.x.toBigInt();
+      let y_bigint = this.y.toBigInt();
+
+      let onCurve =
+        add(mul(x_bigint, mul(x_bigint, x_bigint)), Pallas.b) ===
+        square(y_bigint);
+
+      if (!onCurve) {
+        throw Error(
+          `(x: ${x_bigint}, y: ${y_bigint}) is not a valid group element`
+        );
+      }
+    }
+  }
+
+  // helpers
+  static #fromAffine({ x, y }: { x: bigint; y: bigint; infinity: boolean }) {
+    return new Group({ x, y });
+  }
+
+  static #fromProjective({ x, y, z }: { x: bigint; y: bigint; z: bigint }) {
+    return this.#fromAffine(Pallas.toAffine({ x, y, z }));
+  }
+
+  #toTuple(): [0, FieldVar, FieldVar] {
+    return [0, this.x.value, this.y.value];
+  }
+
+  #isConstant() {
+    return this.x.isConstant() && this.y.isConstant();
+  }
+
+  #toAffine() {
+    return Pallas.fromAffine({
+      x: this.x.toBigInt(),
+      y: this.y.toBigInt(),
+      infinity: false,
+    });
+  }
+
+  /**
+   * Adds this {@link Group} element to another {@link Group} element.
+   *
+   * ```ts
+   * let g1 = Group({ x: 1, y: 1})
+   * let g2 = g1.add(g1)
+   * ```
+   */
+  add(g: Group) {
+    if (this.#isConstant() && g.#isConstant()) {
+      if (this.x.toBigInt() === 0n) {
+        return g;
+      } else if (g.x.toBigInt() === 0n) {
+        return this;
+      } else {
+        let g_proj = Pallas.add(this.#toAffine(), g.#toAffine());
+        return Group.#fromProjective(g_proj);
```

Comment:
see my other comments about consistent treatment of `zero`. This seems to accept `(0, y)` and treats that as zero element. at the same time, adding `g` to `g.neg()` (which should result in zero) returns `(1, 1)` IIUC